---
title: "Introduction to scSGS"
author: Shreyan Gupta, James J Cai
output:
  html_document:
    toc: true
    number_sections: false
    
vignette: >
  %\VignetteIndexEntry{Introduction to scSGS}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
```{r globalOpts, include = FALSE}
options(readr.show_progress = FALSE)
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = ">"
)
```

# Overview 
This tutorial shows how to use **scSGS** for gene function analysis in a WT single-cell RNA seq sample and **Spline-HVG** for highly variable feature selection.
# Installation
scSGS requires [presto](https://github.com/immunogenomics/presto) for fast computation of DE genes using the wilcoxon rank-sum test.
```{r Installation, eval=FALSE}
if(!require(remotes)) install.packages('remotes')

remotes::install_github("immunogenomics/presto")
remotes::install_github("Xenon8778/scSGS")
```

# Import Libraries
For this vignette, letâ€™s use scSGS with the seurat package.
```{r Loading libraries}
library(Seurat)
library(scSGS)
```

# Load datasets
For this tutorial, we will be analyzing the a dataset of Peripheral Blood Mononuclear Cells (PBMC) freely available from 10X Genomics. The raw data can be found [here] (https://cf.10xgenomics.com/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz)
```{r Load Seurat data, warning=FALSE}
# Load the PBMC dataset
pbmc.data <- Read10X(data.dir = "../../scSGS_Extra/data/filtered_gene_bc_matrices/hg19/")
pbmc <- CreateSeuratObject(CreateAssayObject(counts = pbmc.data, project = "pbmc3k", min.cells = 3, min.features = 200))
pbmc
```

## Filtering 
We filter cells that have unique feature counts over 2,500 or less than 200.
```{r Filtering}
pbmc <- subset(pbmc, subset = nFeature_RNA > 200 & nFeature_RNA < 2500)
pbmc
```

# scSGS analysis
We perform scSGS analysis with the most variable gene and find the SG-responsive genes. HVG selection is done using HVG_splinefit which uses each gene's mean expression, CV and dropout rate to compute variability.
```{r scSGS analysis}
SGS_res = scSGS(GetAssayData(pbmc,layer = 'counts'), 'LYZ',nHVG = 500,
                calcHVG = T, verbose = FALSE)

# scSGS DE results
head(SGS_res$DE)
```

```{r HVF data.frame}
# Highly variable genes
head(SGS_res$HVG_df)
```

```{r TFs}
# Known TFs
SGS_res$TF[1:5]
```

# Enrichment analysis
We performed enrichment analysis with the [enrichR](https://cran.r-project.org/web/packages/enrichR/vignettes/enrichR.html) package.
```{r Enrichment, message=FALSE}
library(enrichR)
library(dplyr)
library(ggplot2)

setEnrichrSite("Enrichr")
dbs <- c("KEGG_2021_Human")

# Select top 200 genes
SGS_DE_df = SGS_res$DE %>% arrange(p_val_adj) %>% filter(p_val_adj < 0.01) %>%
  head(200)
DE_gl = SGS_DE_df %>% pull(genes)

# Performing Enrichment with Enrichr
enrichr_df <- enrichr(DE_gl, dbs)

# Plotting
p1 = plotEnrich(enrichr_df[[1]], showTerms = 10, numChar = 100,
                y = "Count", orderBy = "Adjusted.P.value", title = dbs)+
  scale_fill_gradient(low = 'firebrick',high = 'orange')
plot(p1)
```

# sessionInfo
This is the output of `sessionInfo()` on the system on which this document was compiled:
```{r}
date()
sessionInfo()
```
