---
title: "Introduction to scSGS"
author: Shreyan Gupta, James J. Cai
output:
  html_document:
    toc: true
    number_sections: false
    
vignette: >
  %\VignetteIndexEntry{Introduction to scSGS}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
```{r globalOpts, include = FALSE}
options(readr.show_progress = FALSE)
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = ">"
)
```

# Overview 
This tutorial shows how to use **scSGS** to conduct gene function analysis on single-cell RNA-seq data from wild-type samples. We'll also explore the use of **Spline-HVG** for selecting highly variable features.

# Installation
To install the required packages, ensure you have `remotes` installed. Then, install [presto](https://github.com/immunogenomics/presto) for efficient differential expression analysis using the Wilcoxon rank-sum test, and 'scSGS' for gene function analysis.
```{r Installation, eval=FALSE}
if(!require(remotes)) install.packages('remotes')

remotes::install_github("immunogenomics/presto")
remotes::install_github("Xenon8778/scSGS")
```

# Import Libraries
Load the essential libraries for this analysis: 'Seurat' for single-cell data analysis and 'scSGS' for the specific functionalities we'll be using.
```{r Loading libraries, message=FALSE}
library(Seurat)
library(scSGS)
```

# Load datasets
For this tutorial, We'll work with a publicly available Peripheral Blood Mononuclear Cells (PBMC) dataset from 10X Genomics.  The raw data can be found [here] (https://cf.10xgenomics.com/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz). Read the 10X-formatted data and create a Seurat object.
```{r Load Seurat data, warning=FALSE}
# Load the PBMC dataset
pbmc.data <- Read10X(data.dir = "../../scSGS_Extra/data/filtered_gene_bc_matrices/hg19/")
pbmc <- CreateSeuratObject(CreateAssayObject(counts = pbmc.data, project = "pbmc3k", min.cells = 3, min.features = 200))
pbmc
```

## Filtering 
To improve data quality, filter out cells with extremely low or high feature counts, which often indicate low-quality or technical noise. We filter cells that have unique feature counts over 2,500 or less than 200.
```{r Filtering}
pbmc <- subset(pbmc, subset = nFeature_RNA > 200 & nFeature_RNA < 2500)
pbmc
```

# Identification of HVGs
Identify HVGs using the `HVG_splinefit` function. This method leverages a spline-based approach to account for non-linear relationships between mean expression, coefficient of variation, and dropout rate.
```{r}
HVG_res = HVG_splinefit(GetAssayData(pbmc,layer = 'counts'), diptest=TRUE,
                        dropout.filter=TRUE, nHVGs=500)

# HVG results
head(HVG_res[HVG_res$HVG == TRUE,])
```
```{r TFs}
# Highly Variable Genes that are known TFs 
HVG_res[(HVG_res$HVG == TRUE & HVG_res$TF == TRUE),][1:5,]
```
# scSGS analysis
Perform scSGS analysis on the selected HVGs using the `scSGS` function. This function identifies genes that are co-expressed with a specific gene of interest (in this case, 'S100A9').
```{r scSGS analysis}
SGS_res = scSGS(GetAssayData(pbmc, layer='counts'), 'S100A9', nHVG=500,
                rm.mt=T, rm.rp=T,
                calcHVG=TRUE, verbose=TRUE)

# scSGS DE results
head(SGS_res$DE)
```

# Enrichment analysis
Conduct enrichment analysis to gain biological insights into the identified DE genes. Use the [enrichR](https://cran.r-project.org/web/packages/enrichR/vignettes/enrichR.html) package to query various databases for functional annotations.
```{r Enrichment, message=FALSE}
library(enrichR)
library(dplyr)
library(ggplot2)

setEnrichrSite("Enrichr")
dbs <- c("GO_Biological_Process_2023")

# Select top 200 genes
SGS_DE_df = SGS_res$DE %>% arrange(p_val_adj) %>% filter(p_val_adj < 0.01) %>%
  head(200)
DE_gl = SGS_DE_df %>% pull(genes)

# Performing Enrichment with Enrichr
enrichr_df <- enrichr(DE_gl, dbs)

# Plotting
p1 = plotEnrich(enrichr_df[[1]], showTerms = 10, numChar = 60,
                y = "Count", orderBy = "Adjusted.P.value", title = dbs)+
  scale_fill_gradient(low = 'firebrick',high = 'orange')
plot(p1)
```

# sessionInfo
This is the output of `sessionInfo()` on the system on which this document was compiled:
```{r}
date()
sessionInfo()
```
